### PR TITLE
181584868 new transaction details screen design

### PR DESCRIPTION
**Pull request by androidmaven**
Design info here: https://www.notion.so/usehover/Transaction-detail-screen-dc1b5bc8714c45fa8393b59775e28ff4

N.B
"Tapping try again takes the user to the form for that transaction with everything filled out so that the user can edit as necessary (NOT the confirmation screen)." - This is moved to its own separate story.

### CODE REVIEW

**Review by alexosugo — APPROVED** [review]
Code looks good. The view USSD log button can be replaced by a `MaterialButton` that will give the slim feel as in the designs. Probably add some more spacing between the transaction details and make the status icon at the top larger.

File: app/src/main/java/com/hover/stax/transactions/MessagesAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+		return position
+	}
+
+	override fun getItemCount(): Int {
+		return messagesList?.size ?: 0
+	}
+
+	class TransactionMessageViewHolder(var binding: TransactionMessagesItemsBinding) :
+		RecyclerView.ViewHolder(binding.root)
+}
```
Comment (johnswekenyi) [inline app/src/main/java/com/hover/stax/transactions/MessagesAdapter.kt:96]: Please consider adding a line at the end of this file

File: app/src/main/java/com/hover/stax/transactions/TransactionDetailsFragment.kt [hunk 2]
```diff
@@ -31,7 +31,7 @@ import org.json.JSONObject
 import org.koin.androidx.viewmodel.ext.android.viewModel
 
 
-class TransactionDetailsFragment : DialogFragment() {
+class TransactionDetailsFragment : DialogFragment(){
```
Comment (johnswekenyi) [inline app/src/main/java/com/hover/stax/transactions/TransactionDetailsFragment.kt:34]: Please revert this, we need the space

File: app/src/main/java/com/hover/stax/transactions/TransactionDetailsFragment.kt [hunk 3]
```diff
@@ -286,5 +295,4 @@ class TransactionDetailsFragment : DialogFragment() {
             return fragment
         }
     }
-
 }
```
Comment (johnswekenyi) [inline app/src/main/java/com/hover/stax/transactions/TransactionDetailsFragment.kt:298]: Empty line

File: app/src/main/java/com/hover/stax/transactions/TransactionStatus.kt [hunk 4]
```diff
@@ -26,14 +27,32 @@ class TransactionStatus(val transaction: StaxTransaction) {
...
+            HoverAction.AIRTIME -> c.getString(R.string.buy_airtime)
+            HoverAction.P2P -> c.getString(R.string.display_transfer_money, a.to_institution_name)
+            HoverAction.ME2ME -> c.getString(R.string.display_transfer_money, a.to_institution_name)
+            HoverAction.C2B -> c.getString(R.string.display_bill_payment)
+            HoverAction.RECEIVE -> c.getString(R.string.display_money_received)
+            else -> "Other"
+        }.capitalize()
+    }
+
+
```
Comment (johnswekenyi) [inline app/src/main/java/com/hover/stax/transactions/TransactionStatus.kt:55]: Remove this extra line

File: app/src/main/java/com/hover/stax/transactions/USSDLogBottomSheetFragment.kt [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+	companion object {
+		const val UUID = "uuid"
+	}
+
+	override fun onDestroyView() {
+		super.onDestroyView()
+		_binding = null
+	}
+
+}
```
Comment (johnswekenyi) [inline app/src/main/java/com/hover/stax/transactions/USSDLogBottomSheetFragment.kt:79]: Empty line here too

File: app/src/main/res/layout/transaction_details_info_card.xml [hunk 6]
```diff
@@ -18,12 +95,12 @@
                 android:layout_weight="1"
                 android:textAlignment="viewEnd"
                 android:textColor="@color/offWhite"
-                android:textSize="@dimen/text_21" />
```
Comment (johnswekenyi) [inline app/src/main/res/layout/transaction_details_info_card.xml:21]: How do we take care of small device screens with this kind of offset?

File: app/src/main/res/layout/transaction_details_primarystatus.xml [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+        android:layout_margin="@dimen/margin_16"
+        android:id="@+id/viewLogText"
+        android:fontFamily="@font/brutalista_medium"
+        android:textColor="@color/stax_state_blue"
+        android:textStyle="bold"
+        android:padding="@dimen/margin_16"
+        android:background="@drawable/bright_blue_rounded_bg"
+        android:text="@string/viewUSSDLog"/>
+
+</LinearLayout>
```
Comment (johnswekenyi) [inline app/src/main/res/layout/transaction_details_primarystatus.xml:44]: You need an empty line here as well

File: app/src/main/res/layout/transaction_details_secondary_status.xml [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+            android:gravity="center"
+            android:textAlignment="center"
+            android:padding="@dimen/margin_16"
+            android:fontFamily="@font/brutalista_regular"
+            android:textColor="@color/offWhite"
+            android:text="@string/try_again" />
+
+    </RelativeLayout>
+
+</LinearLayout>
```
Comment (johnswekenyi) [inline app/src/main/res/layout/transaction_details_secondary_status.xml:54]: Empty line

File: app/src/main/res/layout/ussd_log_bottomsheet.xml [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+        <androidx.recyclerview.widget.RecyclerView
+            android:id="@+id/sms_recyclerView"
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content"
+            android:layout_marginBottom="@dimen/margin_21"
+            android:layout_marginHorizontal="@dimen/margin_10" />
+
+    </com.hover.stax.views.StaxCardView>
+</androidx.core.widget.NestedScrollView>
```
Comment (johnswekenyi) [inline app/src/main/res/layout/ussd_log_bottomsheet.xml:28]: Empty line

File: app/src/main/res/values/strings.xml [hunk 10]
```diff
@@ -299,12 +299,12 @@
     <string name="noworry_explain">You can review this info in the next step.</string>
     <string name="descrip_multcontacts">%1$d contacts</string>
 
-    <string name="descrip_balance">%1$s check balance</string>
-    <string name="descrip_airtime_sent">%1$s send airtime to %2$s</string>
-    <string name="descrip_transfer_sent">%1$s send money to %2$s</string>
```
Comment (johnswekenyi) [inline app/src/main/res/values/strings.xml:304]: I am not sure I understand this variable `descrip_transfer_sent `

File: app/src/main/res/values/strings.xml [hunk 11]
```diff
@@ -299,12 +299,12 @@
...
+    <string name="descrip_balance">%1$s balance</string>
+    <string name="descrip_airtime_sent">Sent %1$s airtime to %2$s</string>
+    <string name="descrip_transfer_sent">Send %1$s money to %2$s</string>
     <string name="descrip_transfer_received">received money from %1$s</string>
     <string name="descrip_fetch_accounts">fetched accounts from %1$s</string>
-    <string name="descrip_bill_paid">pay %1$s bill</string>
+    <string name="descrip_bill_paid">Paid %1$s to %2$s</string>
     <string name="descrip_recorded">Recorded USSD session on %1$s</string>
     <string name="descrip_airtime_sched">Buy airtime with %1$s for %2$s</string>
```
Comment (johnswekenyi) [inline app/src/main/res/values/strings.xml:309]: Consider making the variable descriptive please

**Review by johnswekenyi — COMMENTED** [review]
LG%C

File: app/src/main/java/com/hover/stax/transactions/TransactionDetailsFragment.kt [hunk 12]
```diff
@@ -202,24 +175,47 @@ class TransactionDetailsFragment : DialogFragment() {
...
         binding.infoCard.detailsStaxUuid.text = transaction.uuid
+        binding.infoCard.detailsStaxStatus.apply {
+            text = transaction.fullStatus.getTitle(requireContext())
+            setCompoundDrawablesWithIntrinsicBounds(0, 0, transaction.fullStatus.getIcon(), 0)
+        }
 
-        setVisibleDetails(transaction)
-        updateStatus(viewModel.action.value, transaction)
+        binding.infoCard.detailsStaxReason.text = transaction.fullStatus.getReason()
+        if(transaction.fee !=null && transaction.fee > 0) binding.infoCard.detailsFee.text = Utils.formatAmount(transaction.fee)
```
Comment (davkutalek) [inline app/src/main/java/com/hover/stax/transactions/TransactionDetailsFragment.kt:203]: Still show fee so long as it is not null even if it is 0

**Review by davkutalek — CHANGES_REQUESTED** [review]
Code looks great, love the bottom sheet for the USSD log.

Couple minor UI fixes: 
- Change the pending icon to the ellipsis, the exclamation is confusing.
- As noted by kombo, USSD log button should be shorter.
- Remove the account icon from the pending message. I think we can leave it for failed, but make it round.
- When successful, the status line is being filled with the balance, it should just say successful.
- There should be a 1 line sized gap after the transaction type. There should be a balance or amount line (depending on transaction type) above the fee line.
- The transaction cost line is showing but not being filled for me on a successful transaction since the fee is 0. If the variable is present but 0 still show it. If the fee is null then hide that line.
- The copy buttons/text should show a ripple when touched. Also use the same UX: either the icon or (tap to copy) is fine but not both.
- Add an X close button to the USSD log bottom sheet https://material.io/components/sheets-bottom#behavior

File: app/src/main/java/com/hover/stax/transactions/TransactionDetailsFragment.kt (hunk 12 above)
Comment (androidmaven) [reply app/src/main/java/com/hover/stax/transactions/TransactionDetailsFragment.kt]: Done

File: app/src/main/res/values-am/strings.xml [hunk 13]
```diff
@@ -279,7 +279,7 @@
 
     <string name="descrip_transfer_received">ገንዘብ ከ %1$s ተቀብሏል</string>
 
-    <string name="descrip_bill_paid">%1$s ሂሳብ ከፍሏል</string>
```
[the remark this answers is not in the document]
Comment (androidmaven) [reply app/src/main/res/values-am/strings.xml:282]: Noted,thanks.

File: app/src/main/java/com/hover/stax/utils/Constants.kt [hunk 14]
```diff
@@ -25,6 +25,8 @@ object Constants {
     const val PERMS_REQ_CODE = 700
 
     const val size55 = 55
```
Comment (davkutalek) [inline app/src/main/java/com/hover/stax/utils/Constants.kt:27]: These should be in dimens.xml

**Review by davkutalek — CHANGES_REQUESTED** [review]
Still some small UI issues with the changes on my device:
- USSD log button hasn't gotten any smaller, but now it is blue text on a slightly lighter blue background. Let's do an (light grey) outlined button with no background color and white text. As per material guidelines I think this is the right way to go. Create a new style for this type of button if necessary.
- The pending details about waiting for an SMS are weirdly aligned, it looks like the spot for the logo in the failed case is invisible instead of gone, should be gone.
- The X for the USSD log is on the right and doesn't do anything. Should be on the left.
- The balance/amount line above the cost line is still missing.
- Transaction id has copy icon while stax id has "(Tap to copy)". Choose one, I think tap to copy is better.

Screenshot_20220414-092023

File: app/src/main/java/com/hover/stax/utils/Constants.kt (hunk 14 above)
Comment (androidmaven) [reply app/src/main/java/com/hover/stax/utils/Constants.kt:27]: The previously added constant is no more in use.

We could probably do a constant cleanup as a chore story.

Comment (androidmaven) [reply app/src/main/java/com/hover/stax/utils/Constants.kt:27]: @davkutalek , thanks for the reviews. Changes effected.

One UI note is unclear: The balance/amount line above the cost line is still missing.
Do you mean this below?

TDscreenshot

Comment (davkutalek) [reply app/src/main/java/com/hover/stax/utils/Constants.kt:27]: @androidmaven yep. The recipient account and to lines should also be below the gap

**Review by davkutalek — CHANGES_REQUESTED** [review]
@androidmaven almost there! The only remaining issue is the spacing and ordering of the items in the list. 

- Pending transactions have lost the gap between type and details. 
- Cost should not show if it is null.
- Success have a giant gap before Stax Id (see screenshot), and no gap after type, and no balance line.
- The details above the gap should be meta info: date, status, reason, sending account, type. The details below the gap should be transaction details: amount/balance, recipient info, fee, ids.
- Instead of "Transaction cost" lets use "[Channel name] transaction fee", eg "MPESA transaction fee" to make it very clear.

**Review by davkutalek — APPROVED** [review]
Everything is super smooth and looks great.

There is still a missing line: "New balance" should always have a line below the fee if balance is not null. Otherwise good to go

**Review by alexosugo — COMMENTED** [review]
Two minor UI issues I found;

- The fees paid to account text appears cut off at the bottom.  
Screenshot_20220420-115926

- Text alignment for the transaction ID is slightly off. 
Screenshot_20220420-115501

This may have been fixed in the other branch so I'll wait to test again.

File: app/src/main/res/values/styles.xml [hunk 15]
```diff
@@ -27,6 +27,16 @@
         <item name="textAllCaps">false</item>
     </style>
 
+    <style name="StaxOutlineButton" parent="Widget.Material3.Button.OutlinedButton">
```
[the remark this answers is not in the document]
Comment (androidmaven) [reply app/src/main/res/values/styles.xml:30]: Thank for catching the UI bugs. They are now fixed in the recent push. 

On the: "Widget.Material3.Button.OutlinedButton", I did try the current material theme as parent, the Material3 actually provided the intended look.

**Review by alexosugo — APPROVED** [review]

**Pull request merged**